### ISSUE TITLE
NameSpacedKey Support?

### ISSUE THREAD

**Issue by mehboss** [CoolLord22/OtherDrops1.13#48]
**Is your feature request related to a problem? Please describe.**
It is not

**Describe the solution you'd like**
I would like integrated support for specifying NameSpacedKey to return an ItemStack. This will allow support for custom recipe / custom item plugins.

**Describe alternatives you've considered**
N/A

**Additional context**
I.E -

ItemDrop
  - requireKey: customrecipes:cursed_sword

Thus, will drop the itemstack cursed sword.
https://www.spigotmc.org/resources/%E2%96%BA%E2%96%BA-custom-crafting-recipes-1-8-x-1-21-x-per-recipe-perms-create-your-own-recipes-%E2%97%84%E2%97%84.36925/

**Comment by CoolLord22** [comment]
Hi! Thank you so much for this feature request! I apologize in advance for the wall of text, did not anticipate the reply being this long haha.

Now unfortunately, I think there’s a slight misunderstanding (unless I am mistaken) about NameSpacedKey and it’s uses. 

The function of NameSpacedKey for the purposes of a custom recipe plugin would be to register with the server the item’s recipe tied to the plugin that is creating said recipe. ~~The only way to access this field from the server is to have an instance of the plugin that registered the NameSpacedKey. This would essentially mean OD would have to then soft depend on any custom recipe plugin that gets requested (since the Main class instance of that plugin would have to be stored in OD to access said recipe).~~

**EDIT: Realized after posting we could just grab a `Plugin` object and parse the user's provided string for a plugin name to get the plugin's instance and then use that for grabbing the recipe from the NameSpacedKey... I'll look into this tomorrow once I have some time!**

The other alternative would be then using the API's of said plugins, but that again creates unnecessary dependencies for plugins that are simply doing the task of allowing the user to creating an item. 

That being said I know OD lacks specific key features (CustomModelData, Banner data, etc) regarding item customization. However, the beta release of OD does add in support for the MythicMobs plugin, which has a very similar item configuration specification system that features full customizability. I believe you should be able to import the item from your hand using `/mm item import <itemName>` and then reference in within OD using `MYTHIC_ITEM@itemName`.

I hate to require you to go use another plugin but since we already have integrations for MM (for not just the item aspects, but for mobs as well) it might be easier for both of us to go with that route. Item comparison and confirmation is already set up for MM items, and it poses another thing to figure out within OD (if that makes any sense)

Closing this for now, but please feel free to re-open if you have any other questions/suggestions for workarounds regarding this. If this is still something you'd like to see in OD I can try to see what I can do, but no promises!

**Comment by CoolLord22** [comment]
While I was sitting here, thought of another solution.

We could consider a complete overhaul on OD's custom items? Or maybe simply add another way to define custom items following a format similar to what's done in MM or customrecipes, where users can define items within a file and reference them throughout the plugin using that custom identifier.

Of course we'd need commands to import/export items from the file, allowing the user to hold an itemstack and have it saved to the plugin. The file would allow for all the custom item datas (model data, durability, bag data, etc) to be specified.

If this is a workaround you'd prefer over using MM, please feel free to reopen this ticket and I'll get started :)

Thanks for using OD! Please reach out if you need anything else :)

**Comment by mehboss** [comment]
I was more or less just looking to add support for your plugin and my plugin, [Custom Recipes.](https://www.spigotmc.org/resources/%E2%96%BA%E2%96%BA-customrecipes-1-8-x-1-21-x-advanced-recipes-made-easy-%E2%97%84%E2%97%84.36925/) The easiest way I could think of was to suggest this:

`NamespacedKey.fromString("customrecipes:" + key);`

Adding a feature in OD to specify a NameSpacedKey option to return a recipe result (ItemStack).
See [here](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/NamespacedKey.html#fromString(java.lang.String)) and [here](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Bukkit.html#getRecipe(org.bukkit.NamespacedKey))

I've had a couple people ask if there was a way that OD could use my custom recipe item results.
I appreciate your quick response! Thank you so much

**Comment by CoolLord22** [comment]
I was under the impression that `NamespacedKey.fromString(String key);` applies a default minecraft namespace (see javadocs below). If this is not the case, that simplifies the itemstack portion significantly! Otherwise, I do believe there is a way to grab the plugin's instance from the name the user provides. Will do some testing and get back to you! 

Once I get the itemstack figured out, all that's left is creating classes so that OD can interpret and compare / give the found itemstacks. Not sure if you use OD yourself or if this was a feature req on your plugin, but if you join our [Discord](https://discordapp.com/invite/eHBxk5q) I can share a working jar with you there once I have it!

---

> Get a NamespacedKey from the supplied string. The default namespace will be Minecraft's (i.e. [minecraft(String)](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/NamespacedKey.html#minecraft(java.lang.String))).

**Comment by mehboss** [comment]
As far as I can tell, as long as you specify the NameSpacedKey with a colon, it will work! For example, my plugin CustomRecipes NameSpace would be "customrecipes:" and you would specify the key of the recipe after (i.e CURSED_SWORD if that is a valid recipe key).

If it does require a plugin's instance, you could definitely check if [Bukkit.getPluginManager().getPlugin(string)](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/plugin/PluginManager.html#getPlugin(java.lang.String)) is null and grab the instance that way :)

Both should be fairly simple

Thank you so much :)
I have joined the discord.

**Comment by CoolLord22** [comment]
Yeah, regardless of what is returned, I think I'll opt toward the user providing the plugin name and namespacekey that way we're not manually hard-coding a string value in ;) this will also allow other plugins that register recipes to be used as well :)

**Comment by mehboss** [comment]
Thank you :)

**Comment by CoolLord22** [comment]
Leaving issue open as:
- ToolAgent portion of Namespace items is not supported
- Issues autoclose once their code changes are pulled into the master branch 

Thanks for the feature req! :)

**Comment by CoolLord22** [comment]
Should now be fully implemented. Custom namespace items can be used as both drops and tool agents now. Formatting for both is `NAMESPACE_ITEM@PluginName:RecipeKey` See example using [CustomRecipes](https://www.spigotmc.org/resources/%E2%96%BA%E2%96%BA-customrecipes-1-8-x-1-21-x-advanced-recipes-made-easy-%E2%97%84%E2%97%84.36925/) below.

```yaml
STONE:
  - trigger: BREAK
    tool: NAMESPACE_ITEM@customrecipes:cursed_pickaxe
    message: "Wow! A sword dropped when you used the cursed pickaxe!"
    drop: NAMESPACE_ITEM@customrecipes:cursed_sword
```
